src/scaladoc/scala/tools/nsc/doc/html/Page.scala

```diff
@@ -13,6 +13,58 @@ import scala.reflect.NameTransformer
 import java.nio.channels.Channels
 import java.io.Writer
 
+object Page {
+  def inlineToStr(inl: Inline): String = inl match {
+    case Chain(items) => items flatMap (inlineToStr(_)) mkString ""
+    case Italic(in) => inlineToStr(in)
+    case Bold(in) => inlineToStr(in)
+    case Underline(in) => inlineToStr(in)
+    case Superscript(in) => inlineToStr(in)
+    case Subscript(in) => inlineToStr(in)
+    case Link(raw, title) => inlineToStr(title)
+    case Monospace(in) => inlineToStr(in)
+    case Text(text) => text
+    case Summary(in) => inlineToStr(in)
+    case HtmlTag(tag) => "<[^>]*>".r.replaceAllIn(tag, "")
+    case EntityLink(in, _) => inlineToStr(in)
+  }
+
+  def templateToPath(tpl: TemplateEntity): List[String] = {
+    def doName(tpl: TemplateEntity): String =
+      (if (tpl.inPackageObject) "package$$" else "") + NameTransformer.encode(tpl.name) + (if (tpl.isObject) "$" else "")
+    def downPacks(pack: TemplateEntity): List[String] =
+      if (pack.isRootPackage) Nil else (doName(pack) :: downPacks(pack.inTemplate))
+    def downInner(nme: String, tpl: TemplateEntity): (String, TemplateEntity) = {
+      if(tpl.inTemplate.isPackage) {
+        (nme + ".html", tpl.inTemplate)
+      } else {
+        downInner(doName(tpl.inTemplate) + "$" + nme, tpl.inTemplate)
+      }
+    }
+    val (file, pack) =
+      if(tpl.isPackage) {
+        ("index.html", tpl)
+      } else {
+        downInner(doName(tpl), tpl)
+      }
+    file :: downPacks(pack)
+  }
+
+  def makeUrl(baseUrl : String, path : List[String]) : String = baseUrl.stripSuffix("/") + "/" + path.reverse.mkString("/")
```

Comment:
I recommend doing all URI-related computations strongly typed with `java.net.URI` instead of stringly typed to ensure that paths are concatenated correctly and to make it clear what has been encoded and what has not.

I did not look at the rest of the scaladoc code to see how pervasive this kind of URI handling is. If it is already done with string manipulation in many cases a rewrite would best be left for another PR.